import {
  NextFunction,
  Request,
  Response,
} from 'express';
import {
  NoteService,
  UserService,
} from '../../services';
import { ERROR_CODES } from '../../constants';
import { Fail } from '../../helpers';
import { compareSync } from 'bcryptjs';

/**
 * GET /api/v1/user
 */
export const getUser = async (req: Request, res: Response , next: NextFunction) => {
  try {
    const userService = new UserService(req.logger, req.mongo);

    const { id } = req.user;
    const { options } = req.mquery;

    const { user } = await userService.fetchUser({ id }, options)
      .catch((error: any) => {
        throw error;
      });
    delete user.password;

    res.jsend.success({ user });
  } catch (error) {
    next(error);
  }
};

/**
 * PUT /api/v1/user
 */
export const putUser = async (req: Request, res: Response , next: NextFunction) => {
  try {
    const userService = new UserService(req.logger, req.mongo);

    const { id } = req.user;
    const {
      username,
      email,
      name,
    } = req.body;

    await userService.updateUser({ id }, {
      username,
      email,
      name,
    })
      .catch((error: any) => {
        throw error;
      });

    res.jsend.success(undefined, 204);
  } catch (error) {
    next(error);
  }
};

/**
 * DELETE /api/v1/user
 */
export const deleteUser = async (req: Request, res: Response , next: NextFunction) => {
  try {
    const userService = new UserService(req.logger, req.mongo);
    const noteService = new NoteService(req.logger, req.mongo);

    const { id } = req.user;
    const { password } = req.body;

    const { user } = await userService.fetchUser({ id }, { projection: { password: 1 } })
      .catch((error: any) => {
        throw error;
      });
    const isMatch = compareSync(password, user.password);
    if (!isMatch) {
      throw new Fail(403)
        .error({
          errorCode: ERROR_CODES.INVALID,
          keys: ['password'],
          message: 'Provided password is invalid.',
        })
        .build();
    }
    await userService.deleteUser({ id })
      .catch((error: any) => {
        throw error;
      });
    await noteService.deleteNotes({ id })
      .catch((error: any) => {
        throw error;
      });

    res.jsend.success(undefined, 204);
  } catch (error) {
    next(error);
  }
};


/**
 * PUT /api/v1/user/password
 */
export const putUserPassword = async (req: Request, res: Response , next: NextFunction) => {
  try {
    const userService = new UserService(req.logger, req.mongo);

    const { id } = req.user;
    const { currentPassword, newPassword } = req.body;

    const { user } = await userService.fetchUser({ id }, { projection: { password: 1 } })
      .catch((error: any) => {
        throw error;
      });
    const isMatch = compareSync(currentPassword, user.password);
    if (!isMatch) {
      throw new Fail(403)
        .error({
          errorCode: ERROR_CODES.INVALID,
          keys: ['currentPassword'],
          message: 'Provided current password is invalid.',
        })
        .build();
    }
    await userService.updateUser({ id }, { password: newPassword })
      .catch((error: any) => {
        throw error;
      });

    res.jsend.success(undefined, 204);
  } catch (error) {
    next(error);
  }
};
